Add Bitcoin.de exchange and ICONOMI platform

coinyon opened the pull request:
Closes #991 
Closes #629

coinyon commented:
This is a first version that requires some cleanup and tests.

LefterisJP commented:
@coinyon any update here?

coinyon commented:
I don't have too much time on my hands currently, but I'm rebasing this regularly and will fix the CI issues. I'm let you know when this is ready from my view

coinyon commented:
@LefterisJP All checks pass now. How to proceed from here?

- This currently supports balances and trades for both Bitcoin.de and ICONOMI.
- ICONOMI funds (except for BLX) are not supported and ignored.

LefterisJP commented:
Hey @coinyon thanks a lot! I would have to just check it myself, inspect the code, perhaps get someone to give me an API key for testing and see it works and then we can merge in develop after v1.6.2 so that it goes for v.1.7.0

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,250 @@
+
+    timestamp = raw_trade['timestamp']
+
+    if raw_trade['type'] == 'buy_fund' \
```
can you please change this? In the code we avoid splitting lines with `\`

Instead do it like below:

```python
unknown asset = (
    raw_trade['type'] == 'buy_fund' and
    raw_trade['target_ticker'] not in SUPPORTED_FUND_TICKERS
)
if unknown_asset:
    ...
```

Both here and in the rest of the places where you split the line like this below.

coinyon replied:
Is this ok?

```
if (                                                                                                         
    raw_trade['type'] == 'buy_fund' and                                                                      
    raw_trade['target_ticker'] not in SUPPORTED_FUND_TICKERS                                                 
):                                                                                                           
    raise UnknownAsset(raw_trade['target_ticker'])
```

LefterisJP replied:
yeah should be. The only reason for the extra variable is readability.

LefterisJP commented on rotkehlchen/exchanges/manager.py:
```diff
@@ -22,6 +22,7 @@
     'coinbase',
     'coinbasepro',
     'gemini',
+    'bitcoinde',
```
Shouldn't `'iconomi'` also be here?

coinyon replied:
Yep. Changed.

LefterisJP commented on rotkehlchen/exchanges/bitcoinde.py:
```diff
@@ -0,0 +1,249 @@
+        self.msg_aggregator = msg_aggregator
+
+    def _generate_signature(self, request_type: str, url: str, nonce: str) -> str:
+        md5_empty = 'd41d8cd98f00b204e9800998ecf8427e'  # this is md5('')
```
Wouldn't it make sense to move this to the top of the file as a constant?

LefterisJP commented on rotkehlchen/exchanges/bitcoinde.py:
```diff
@@ -0,0 +1,249 @@
+
+    def _api_query(
+            self,
+            verb: str,
```
you can actually even enforce this with types and mypy. Example: https://github.com/rotki/rotki/blob/4500381d5c527d255256f1a32d8261722ce028c2/rotkehlchen/exchanges/bittrex.py#L206

LefterisJP commented on rotkehlchen/exchanges/bitcoinde.py:
```diff
@@ -0,0 +1,249 @@
+            'x-api-nonce': nonce,
+        })
+        if data != '':
+            self.session.headers.update({
```
There may be a bug hidden here. We had it in some other exchange and @vnavascues spotted it.

Since `self.session.headers` keeps state in between calls, a call may have headers from a previous api call unless it's cleaned.

Depending on the headers and the way the Remote server behaves this may end up in an error.

For example here imagine the following scenario:

1. A call with data of length 100 happens succesfully.
2. Then you do a call without data, so we never get in here, but the headers of the previous session are still active. Bitcoin.de server may process this incorrectly.

I guess the safest approach is to determine which headers should stay session-wide, and populate them in the beginning (the exchange constructor). And then in each call append the call specific headers.

Note: This buggy behaviour may still be present in some of our exchange integrations. @vnavascues fixed it for one or two but he also wanted to check the others to make sure.

coinyon replied:
Moved all headers to an argument to `get` as they are request-specific

LefterisJP commented on rotkehlchen/exchanges/bitcoinde.py:
```diff
@@ -0,0 +1,249 @@
+        return json_ret
+
+    def validate_api_key(self) -> Tuple[bool, str]:
+        "Validates that the Bitcoin.de API key is good for usage in Rotki"
```
Can you put triple quotes here? I am surprised no linter hit here. It's how we have all our method docstrings.

LefterisJP commented on rotkehlchen/exchanges/bitcoinde.py:
```diff
@@ -0,0 +1,249 @@
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+    ) -> List:
```
Why not specify it as `List[Trade]` here?

LefterisJP commented on rotkehlchen/exchanges/bitcoinde.py:
```diff
@@ -0,0 +1,249 @@
+
+            if tx['state'] != 1:
+                continue
+            if timestamp and timestamp < start_ts:
```
Why `if timestamp` here?

Is it just to silence mypy? You can probably do it by putting a `timestamp: Timestamp` above the for. If not just use` type: ignore comment`

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,260 @@
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+SUPPORTED_FUND_TICKERS = ('BLX',)
```
So only one asset is supported for iconomi? I have not really used it. What is a fund ticker?

coinyon replied:
I removed support for ICONOMI funds/strategies. Currently only single assets are supported.

BLX worked in previous versions of Rotki because it was tokenized as an ERC20 as well. But since Rotki removed support for BLX, I removed it from the importer as well.

coinyon replied:
In ICONOMI all funds/strategies had tickers and they wanted to tokenize them as well. But I think they no longer planning to do that.

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,260 @@
+    def _api_query(
+            self,
+            verb: str,
+            path: str,
```
same comment as in bitcoinde

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,260 @@
+            'ICN-TIMESTAMP': timestamp,
+        })
+        if data != '':
+            self.session.headers.update({
```
same comment for the potential sessions bug as in bitcoinde

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,260 @@
+
+        try:
+            response = getattr(self.session, verb)(request_url, data=data, timeout=30)
+        except ReadTimeout as exc:
```
Why only ReadTimeout? I would suggest catching all possible request exceptions with `requests.exceptions.RequestException`, logging an error and as you already do raise a RemoteError.

Essentially same thing we do in other exchanges. Example:

https://github.com/rotki/rotki/blob/4500381d5c527d255256f1a32d8261722ce028c2/rotkehlchen/exchanges/bittrex.py#L288-L295

LefterisJP commented on rotkehlchen/exchanges/bitcoinde.py:
```diff
@@ -0,0 +1,249 @@
+
+        log.debug('Bitcoin.de API Query', verb=verb, request_url=request_url)
+
+        response = getattr(self.session, verb)(request_url, data=data)
```
You should put this in a try/except `requests.exceptions.RequestException` and log it and raise `RemoteError`

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,260 @@
+        return json_ret
+
+    def validate_api_key(self) -> Tuple[bool, str]:
+        "Validates that the ICONOMI API key is good for usage in Rotki"
```
Use triple `"""`

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,260 @@
+                    f' Ignoring its balance query.',
+                )
+
+        return (balances, "")
```
no need for the parentheses, also use single quotes.

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,260 @@
+
+    def query_balances(self, **kwargs: Any) -> Tuple[Optional[Dict[Asset, Dict[str, Any]]], str]:
+        balances = {}
+        resp_info = self._api_query('get', 'user/balance')
```
Here, just like in other exchange `query_balances()` put this in try/except and catch `RemoteError`. If there is an error log it and return `None, error_msg`.

Example:

https://github.com/rotki/rotki/blob/4500381d5c527d255256f1a32d8261722ce028c2/rotkehlchen/exchanges/bittrex.py#L306-L315

LefterisJP commented on rotkehlchen/exchanges/bitcoinde.py:
```diff
@@ -0,0 +1,249 @@
+                'usd_value': balance['total_amount'] * usd_price,
+            }
+
+        return balances, ""
```
please use single quote `''` here

LefterisJP commented on rotkehlchen/exchanges/bitcoinde.py:
```diff
@@ -0,0 +1,249 @@
+
+    def query_balances(self, **kwargs: Any) -> Tuple[Optional[Dict[Asset, Dict[str, Any]]], str]:
+        balances = {}
+        resp_info = self._api_query('get', 'account')
```
Put this in try/except and catch the RemoteError. Then if there is a RemoteError, catch it, log it and return `None, error_message` just like we do in other exchanges.

Example:

https://github.com/rotki/rotki/blob/4500381d5c527d255256f1a32d8261722ce028c2/rotkehlchen/exchanges/bittrex.py#L306-L315

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,260 @@
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+    ) -> List:
```
List[Trade]

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,260 @@
+        trades = []
+        for tx in all_transactions:
+            timestamp = tx['timestamp']
+            if timestamp and timestamp < start_ts:
```
why `if timestamp` here? When can it be `None`?

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,260 @@
+                try:
+                    trades.append(trade_from_iconomi(tx))
+                except UnknownAsset:
+                    log.warning('Ignoring transaction %s because of unsupported asset', str(tx))
```
Please use f-string notation (for consistency's sake with the rest of the codebase)

LefterisJP reviewed (commented):
Hey @coinyon thanks a lot. So I reviewed the code and left some comments.

I guess apart from the tests you also tested it with real data that I assume you have for both exchanges right?

It's not needed for this PR and maybe you can add in another one ... but if you do here its' also great.

If the API of each exchange allows, it would be cool to have a `"make_sure_assets_are_known"` test.

For example check binance's one here:

https://github.com/rotki/rotki/blob/develop/rotkehlchen/tests/exchanges/test_binance.py#L163-L191

What it does it query the exchange's supported assets (so a requirement is for the exchange to have an endpoint to query what assets it supports), goes through them and generates a test warning for each asset Rotki does not yet know.

This way each time an exchange adds an asset, we see it from the warnings and also add it to Rotki.

coinyon commented:
Yea, both importers are tested with real data and have been in my use for some months now :-) Only in buchfink tho.

Today,

* I fixed all the issues that came up in the review,
* I added missing asset detection in tests for live queried data for ICONOMI,
* I added a constant list of Bitcoin.de pairs that is checked in tests. Bitcoin.de has no API for getting all supported tickers.

There are two problems with the ICONOMI API (not on our side):

1. Invalid credentials will trigger a ReadTimeout. I will get in contact with the ICONOMI team about this.
2. Values in balances are said to be returned in USD, but are actually EUR. I will also ask the team. For now I used Inquirer.

LefterisJP commented on rotkehlchen/exchanges/iconomi.py:
```diff
@@ -0,0 +1,287 @@
+
+                # There seems to be a bug in the ICONOMI API regarding balance_info['value'].
+                # The value is supposed to be in USD, but is actually returned
+                # in EUR. So let's use the Inquirer for now.
```
This is really interesting. Would be nice to see if their support can say anything about this.

LefterisJP commented:
Hey @coinyon thanks a lot!

> I added a constant list of Bitcoin.de pairs that is checked in tests. Bitcoin.de has no API for getting all supported tickers.

For that can you be checking it from time to time if they add anything new? Or is there some blog they post such news?

> There are two problems with the ICONOMI API (not on our side):

Interesting. I saw the EUR thing. The read timeout is really strange. Let's see what their support responds.

LefterisJP reviewed (approved):
Hey @coinyon so it looks good to me and the single failing test is unrelated.

Can you add two changelog entries also before we merge?
